introducing-pgzx.mdx

```diff
@@ -0,0 +1,221 @@
+
+A major reason to use Zig for Postgres development is its ability to interoperate with C code. Zig supports the C ABI, can work with C pointers and data types, including NULL terminated strings, and can even translate C header files into Zig code. While the automatic translating of C macros to Zig is not perfect yet, it can still help a great deal. This makes Zig a great choice for working with very large C code bases, like Postgres happens to be.
+
+## A walkthrough of pgzx
+
+Because Zig can call any C function and translate C macros to inline Zig functions, you can write Postgres extensions in Zig without any additional tooling by following the same steps you would for a C extension. However, in practice, a framework like `pgzx` simplifies the process significantly by providing a development environment, a set of utilities and wrappers for Postgres APIs, common error handling, and more.
+
+### A sample extension
+
+`pgzx` has at the moment 2 sample extensions that you can take inspiration from. [char_count_zig](https://github.com/xataio/pgzx/tree/main/examples/char_count_zig) is a minimal extension, while [pg_audit_zig](https://github.com/xataio/pgzx/tree/main/examples/pgaudit_zig) is more complex and shows more of the features of pgzx.
+
+Let's look at the `char_count_zig`, which is only slightly more than a "Hello, World!" of Postgres extensions. It adds a functions that counts how many times a character appears in a string. This was inspired by [this tutorial](https://www.highgo.ca/2019/10/01/a-guide-to-create-user-defined-extension-modules-to-postgres/), which shows how to do this in plpqsql and C.
+
+```sql
+postgres=# select char_count_zig('hi hii', 'i');
+
+ char_count_zig
+----------------
+              3
+(1 row)
+```
+
+Here is the Zig code:
+
+```zig
+const std = @import("std");
+const pgzx = @import("pgzx");
+
+comptime {
+    pgzx.PG_MODULE_MAGIC();
+
+    pgzx.PG_FUNCTION_V1("char_count_zig", char_count_zig);
+}
+
+fn char_count_zig(input_text: []const u8, target_char: []const u8) !u32 {
+    if (target_char.len > 1) {
+        return pgzx.elog.Error(@src(), "Target char is more than one byte", .{});
+    }
+
+    pgzx.elog.Info(@src(), "input_text: {s}\n", .{input_text});
+    pgzx.elog.Info(@src(), "target_char: {s}\n", .{target_char});
+    pgzx.elog.Info(@src(), "Target char len: {}\n", .{target_char.len});
+
+    var count: u32 = 0;
+    for (input_text) |char| {
+        if (char == target_char[0]) {
+            count += 1;
+        }
+    }
+    return count;
+}
+```
+
+If you compare it with the [C version of the same extension](https://gist.github.com/tsg/2ddbf7ecff7d13e8a8bf92b6b8c5e7fa), you will notice it's quite similar, but the Zig version is actually a bit more concise. This is partially because Zig is more expressive, but also because `pgzx` does a bit more of the work.
```

Comment:
It might be easier to compare if you put the C code in a `<TabbedCode>` block alongside the zig one, rather than use a gist.

Reply:
I ended up making a `CompareCode` component for this.

https://github.com/xataio/mdx-blog/pull/241
https://github.com/xataio/frontend-next/pull/3321

Reply:
Looks great, thanks!